File: common/java/src/main/java/org/aiddl/common/math/linear_algebra/LupDecomposition.java
package org.aiddl.common.math.linear_algebra;

import java.util.ArrayList;
import java.util.List;

import org.aiddl.core.interfaces.Function;
import org.aiddl.core.representation.Term;

public class LupDecomposition implements Function {

	@Override
	public Term apply(Term A_in) {
		if ( A_in.size() != A_in.get(0).size() ) {
			throw new IllegalArgumentException(String.format("Cannot perform LUP decomposition on %dx%d matrix (format must be MxM)", A_in.size(), A_in.get(0).size()));
		}
		
		int n = A_in.size();
		List<Integer> pi = new ArrayList<>();
		List<List<Double>> A = new ArrayList<>();
		for ( int i = 0 ; i < n ; i++ ) {
			pi.add(i);
			List<Double> A_row = new ArrayList<>();
			for ( int j = 0 ; j < n ; j++ ) {
				A_row.add(A_in.get(i).get(j).getDoubleValue());
			}
			A.add(A_row);
		}
		
		for ( int k = 0 ; k < n ; k++ ) {
			double p = 0.0;
			Integer k_d = null;
			for ( int i = k ; i < n ; i++ ) {
				if ( Math.abs(A.get(i).get(k)) > p ) {
					p = Math.abs(A.get(i).get(k));
					k_d = i;
				}
			}
			if ( p == 0.0 ) {
				throw new IllegalStateException("Matrix is singular.");
			}
			{
				int tmp = pi.get(k);
				pi.set(k, pi.get(k_d));
				pi.set(k_d, tmp);
			}
			for ( int i = 0 ; i < n ; i++ ) {
				Double tmp = A.get(k).get(i);
				A.get(k).set(i, A.get(k_d).get(i));
				A.get(k_d).set(i, tmp);
			}
			for ( int i = k+1 ; i < n ; i++ ) {
				A.get(i).set(k, A.get(i).get(k)/A.get(k).get(k));
				for ( int j = k+1 ; j < n ; j++ ) {
					Double a_ij = A.get(i).get(j);
					Double a_kj = A.get(k).get(j);
					A.get(i).set(j, a_ij - A.get(i).get(k)*a_kj);					
				}
			}
		}
		
		List<Term> L = new ArrayList<>();
		List<Term> U = new ArrayList<>();
		for ( int i = 0 ; i < n ; i++ ) {
			List<Term> L_row = new ArrayList<>();
			List<Term> U_row = new ArrayList<>();
			for ( int j = 0 ; j < n ; j++ ) {
				if ( i > j ) {
					L_row.add(Term.real(A.get(i).get(j)));
					U_row.add(Term.real(0.0));
				} else if ( i == j ) {
					L_row.add(Term.real(1.0));
					U_row.add(Term.real(A.get(i).get(j)));
				} else {
					L_row.add(Term.real(0.0));
					U_row.add(Term.real(A.get(i).get(j)));
				}
			}
			L.add(Term.tuple(L_row));
			U.add(Term.tuple(U_row));
		}
				 
		List<Term> pi_r = new ArrayList<>();
		for ( int k : pi ) {
			pi_r.add(Term.integer(k));
		}
		
		return Term.tuple(Term.tuple(L), Term.tuple(U), Term.tuple(pi_r));
	}

}

File: common/scala/src/main/scala/org/aiddl/common/scala/math/linear_algebra/LupDecomposition.scala
package org.aiddl.common.scala.math.linear_algebra

import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.representation.*

import scala.collection.mutable.HashMap

import scala.collection.mutable

class LupDecomposition extends Function {
  def apply( x: Term ): Term = ???

  def apply( a: Matrix ): (Matrix, Matrix, Matrix) = {
    require( a.m == a.n )
    val pi = Array.range(0, a.m).map(i => Num(i))
    val a_mut: mutable.Map[(Int, Int), Num] = new HashMap[(Int, Int), Num]().withDefault((i: Int , j: Int) => a(i, j).asNum)
    (0 until a.m).foreach( k => {
      val k_d = (k until a.m).maxBy( i => a_mut(i, k).abs )
      if ( a_mut(k_d, k) == Num(0) ) throw new IllegalArgumentException("Matrix is singular.")

      // fix permunation
      val tmp = pi(k)
      pi.update(k, pi(k_d))
      pi.update(k_d, tmp)
      // swap rows
      (0 until a.m).foreach( i => {
        val tmp = a_mut(k, i)
        a_mut.update((k, i), a_mut(k_d, i))
        a_mut.update((k_d, i), tmp)
      })
      (k+1 until a.m).foreach( i => {
        a_mut.update((i, k), a_mut(i, k) / a_mut(k, k)  )
        (k+1 until a.n).foreach( j => {
          a_mut.update((i, j), a_mut(i, j) - a_mut(i, k) * a_mut(k, j))
        })
      })
    })

    val L = Tuple((0 until a.m).map( i => {
      Tuple((0 until a.m).map( j =>
        if ( i > j ) a_mut(i, j) else if (i == j) Num(1.0) else Num(0.0) ): _*)}): _*)
    val U = Tuple((0 until a.m).map( i => {
      Tuple((0 until a.m).map( j =>
        if ( i > j ) Num(0.0) else a_mut(i, j) ): _*)}): _*)

    (AiddlMatrix(L), AiddlMatrix(U), AiddlMatrix.vec(pi: _*))
  }
}
